ml-agents/mlagents/trainers/policy/checkpoint_manager.py

```diff
@@ -0,0 +1,95 @@
+    def remove_checkpoint(checkpoint: Dict[str, Any]) -> None:
+        """
+        Removes a checkpoint stored in checkpoint_list.
+        If checkpoint cannot be found, no action is done.
+        :param checkpoint: A checkpoint stored in checkpoint_list
+        """
+        file_path: str = checkpoint["file_path"]
+        if os.path.exists(file_path):
+            os.remove(file_path)
+            logger.info(f"Removed checkpoint model {file_path}.")
+        else:
+            logger.info(f"Checkpoint at {file_path} could not be found.")
+        return
+
+    @classmethod
+    def manage_checkpoint_list(
+        cls, behavior_name: str, keep_checkpoints: int
+    ) -> List[Dict[str, Any]]:
+        """
+        Ensures that the number of checkpoints stored are within the number
+        of checkpoints the user defines. If the limit is hit, checkpoints are
+        removed to create room for the next checkpoint to be inserted.
+        :param category: The category (usually behavior name) of the parameter.
+        :param keep_checkpoints: Number of checkpoints to record (user-defined).
+        """
+        checkpoints = cls.get_checkpoints(behavior_name)
+        while len(checkpoints) >= keep_checkpoints:
+            if (keep_checkpoints <= 0) or (len(checkpoints) == 0):
+                break
+            CheckpointManager.remove_checkpoint(checkpoints.pop(0))
+        return checkpoints
+
+    @classmethod
+    def track_checkpoint_info(
+        cls, behavior_name: str, new_checkpoint: Checkpoint, keep_checkpoints: int
+    ) -> None:
+        """
+        Make room for new checkpoint if needed and insert new checkpoint information.
+        :param category: The category (usually behavior name) of the parameter.
+        :param value: The new checkpoint to be recorded.
+        :param keep_checkpoints: Number of checkpoints to record (user-defined).
+        """
+        checkpoints = cls.manage_checkpoint_list(behavior_name, keep_checkpoints)
+        new_checkpoint_dict = attr.asdict(new_checkpoint)
+        checkpoints.append(new_checkpoint_dict)
+
+    @classmethod
+    def track_final_model_info(
+        cls, behavior_name: str, final_model: Checkpoint, keep_checkpoints: int
+    ) -> None:
+        """
+        Ensures number of checkpoints stored is within the max number of checkpoints
+        defined by the user and finally stores the information about the final
+        model (or intermediate model if training is interrupted).
+        :param category: The category (usually behavior name) of the parameter.
+        :param final_model_path: The file path of the final model.
+        :param keep_checkpoints: Number of checkpoints to record (user-defined).
+        """
+        CheckpointManager.manage_checkpoint_list(behavior_name, keep_checkpoints)
+        final_model_dict = attr.asdict(final_model)
+        GlobalTrainingStatus.set_parameter_state(
+            behavior_name, StatusType.FINAL_MODEL, final_model_dict
```

Comment:
I do not think we need a FINAL_MODEL StatusType.
The final model should simply be the last in the checkpoint in StatusType.CHECKPOINTS.
If you really want to keep it rename FINAL_MODEL to FINAL_CHECKPOINT since it is not a model.
You could also rename to LATEST_CHECKPOINT and modify it in `track_checkpoint_info` every time a model is added.

Reply:
I think from the user's perspective there is a difference between the resulting model and a checkpoint.  This is actually referring to the model output at the end of training, when max_steps is reached.  Maybe it would be clearer to rename the `Checkpoint` class to `ModelInfo` or similar?

Reply:
Checkpoint might be confused with a TensorFlow checkpoint.
> This is actually referring to the model output at the end of training, when max_steps is reached.

I understand this, but I think what should happen at the end of training is that a new checkpoint/model info should be stored and then the last checkpoint/model info's `.nn` file is copied and given to the user at the end of training.